sklearn/metrics/ranking.py

```diff
@@ -346,13 +367,113 @@ def _binary_roc_auc_score(y_true, y_score, sample_weight=None):
+            if len(unique_labels) != len(labels):
+                raise ValueError("Parameter 'labels' must be unique")
+            if len(unique_labels) != y_score.shape[1]:
+                raise ValueError(
+                    "Number of given labels, {0}, not equal to the number "
+                    "of columns in 'y_score', {1}".format(
+                        len(unique_labels), y_score.shape[1]))
+            if set(np.unique(y_true)) > set(unique_labels):
+                raise ValueError(
+                    "'y_true' contains labels not in parameter 'labels'")
+        if multiclass == "ovo":
+            if sample_weight is not None:
+                raise ValueError("Parameter 'sample_weight' is not supported"
+                                 " for multiclass one-vs-one ROC AUC."
+                                 " 'sample_weight' must be None in this case.")
+            y_true_encoded = _encode_y_true_multiclass_ovo(
+                y_true, y_score, labels)
+            # Hand & Till (2001) implementation (ovo)
+            return _average_multiclass_ovo_score(
+                _binary_roc_auc_score,
+                y_true_encoded,
+                y_score,
+                average=average)
+        else:
+            # ovr is same as multi-label
+            if labels is None:
+                labels = np.unique(y_true)
+            y_true_multilabel = label_binarize(y_true, labels)
+            return _average_binary_score(
+                 _binary_roc_auc_score, y_true_multilabel, y_score, average,
+                 sample_weight=sample_weight)
+    elif y_type == "binary":
         labels = np.unique(y_true)
         y_true = label_binarize(y_true, labels)[:, 0]
+        return _average_binary_score(
+            _binary_roc_auc_score, y_true, y_score, average,
+            sample_weight=sample_weight)
+    else:
+        return _average_binary_score(
+            _binary_roc_auc_score, y_true, y_score, average,
+            sample_weight=sample_weight)
+
+
+def _encode_y_true_multiclass_ovo(y_true, y_score, labels):
+    """Encodes y_true for multiclass scoring where y_score is a probability
+    matrix
+
+    Parameters
+    ----------
+    y_true : numpy array, shape = (n_samples, )
+        True multiclass labels
+
+    y_score : numpy array, shape = (n_samples, n_classes)
+        Target scores corresponding to probability estimates of a sample
+        belonging to a particular class
+
+    labels : array-like, shape = (n_classes, ) or None
+        List of labels to index ``y_score`` used. If ``None``,
+        the lexicon order of ``y_true`` is used to index ``y_score``.
+
+    Returns
+    -------
+    y_true_encoded : numpy array, shape = (n_samples, )
+        Encoded y_true
+    """
+    if labels is not None:
+        y_true_encoded = np.empty_like(y_true, dtype=np.int32)
+        for i, label in enumerate(labels):
+            y_true_encoded[y_true == label] = i
```

Comment:
@thomasjpfan 

> In this context, the `labels` may be an unordered list, where the positions represents different columns in `y_score`

I'm not really sure why we need to handle the unordered case... but if we are, perhaps we should at least leave a note in `sklearn/preprocessing/label.py` that if its encoders are extended to handle unsorted `uniques`, then we should remove this function... actually `_encode_python` appears to handle unsorted `uniques`?

Reply:
It does look like `_encode_python` handles the unordered case.

The purpose of `labels` was to support use case where `y_true` are strings and `y_score`'s columns are not labeled in lexicon order.

If we do not handle the unordered case, it would be better to remove the `labels` keyword.